flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/InternalTimeServiceManagerImpl.java

```diff
@@ -99,23 +101,30 @@ private InternalTimeServiceManagerImpl(
      * <p><b>IMPORTANT:</b> Keep in sync with {@link InternalTimeServiceManager.Provider}.
      */
     public static <K> InternalTimeServiceManagerImpl<K> create(
-            CheckpointableKeyedStateBackend<K> keyedStatedBackend,
+            CheckpointableKeyedStateBackend<K> keyedStateBackend,
             ClassLoader userClassloader,
             KeyContext keyContext,
             ProcessingTimeService processingTimeService,
             Iterable<KeyGroupStatePartitionStreamProvider> rawKeyedStates)
             throws Exception {
-        final KeyGroupRange keyGroupRange = keyedStatedBackend.getKeyGroupRange();
+        final KeyGroupRange keyGroupRange = keyedStateBackend.getKeyGroupRange();
+
+        KeyedStateBackend<?> rootKeyedStateBackend =
+                keyedStateBackend instanceof ProxyKeyedStateBackend
+                        ? ((ProxyKeyedStateBackend<?>) keyedStateBackend)
+                                .getProxiedKeyedStateBackend()
+                        : keyedStateBackend;
+
         final boolean requiresSnapshotLegacyTimers =
-                keyedStatedBackend instanceof AbstractKeyedStateBackend
-                        && ((AbstractKeyedStateBackend<K>) keyedStatedBackend)
+                rootKeyedStateBackend instanceof AbstractKeyedStateBackend
+                        && ((AbstractKeyedStateBackend<K>) rootKeyedStateBackend)
```

Comment:
WDYT about pulling `requiresLegacySynchronousTimerSnapshots` to the interface?
Then, with `unwrap` added we could just call
`rootKeyedStateBackend.unwrap().requiresLegacySynchronousTimerSnapshots()`.

Reply:
Yes, that's a good idea